# Real-world Data Wrangling

In this project, you will apply the skills you acquired in the course to gather and wrangle real-world data with two datasets of your choice.

You will retrieve and extract the data, assess the data programmatically and visually, accross elements of data quality and structure, and implement a cleaning strategy for the data. You will then store the updated data into your selected database/data store, combine the data, and answer a research question with the datasets.

Throughout the process, you are expected to:

1. Explain your decisions towards methods used for gathering, assessing, cleaning, storing, and answering the research question
2. Write code comments so your code is more readable

Before you start, install the some of the required packages. 

# Research Question 

### How are housing price and Homelessness attributes correlated? 


## Definitions


df = This is the first data, from kaggle that Have been scrapped from Realtor website. 

df2 = This is the second data, The Homelessness, in the period 2007 - 2016 - Kaggle. 


In [5]:
!python -m pip install kaggle==1.6.12

### **1.1.** Problem Statement
In 2-4 sentences, explain the kind of problem you want to look at and the datasets you will be wrangling for this project.

>I would like to work on Real estate in USA

In [7]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import requests
import zipfile
import json
import requests
import os



## 1. Gather data

In this section, you will extract data using two different data gathering methods and combine the data. Use at least two different types of data-gathering methods.

### **1.2.** Gather at least two datasets using two different data gathering methods

## Dataset 1 
The first data set is a spreadsheet a full of informations about the USA housing data, prices, places and dates and of sold or not, I downlowded it manully from Kaggle, and this data was collected from https://www.realtor.com/ - A real estate listing website operated by the News Corp subsidiary Move, Inc. and based in Santa Clara, California. 
I choosed this dataset to study the housing market in USA. Download it manually. 


In [10]:
df = pd.read_csv("realtor-data.zip.csv")


## Dataset 2

The Seconed Dataset is about homelessness is USA it has 2 csv files, homlessness in the period 2007 - 2016 and population csv. 
These raw data sets contain Point-in-Time (PIT) estimates and national PIT estimates of homelessness as well as national estimates of homelessness by state and estimates of chronic homelessness from 2007 - 2016. Estimates of homeless veterans are also included beginning in 2011. The accompanying Housing Inventory Count (HIC) data is available as well from 2007 - 2016." (Department of Housing and Urban Development).
> Type: zip File, CSV File

> Method: The data was gathered using the "API" method from Kaggle.
> Dataset variables:
Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', count

omes in $1000's

In [12]:
import kaggle

In [13]:
from kaggle.api.kaggle_api_extended import KaggleApi


In [14]:
api = KaggleApi()
api.authenticate()
api.dataset_download_files("adamschroeder/homelessness",
                           "/2007-2016-Homelessnewss-USA.csv")

Dataset URL: https://www.kaggle.com/datasets/adamschroeder/homelessness


In [161]:

zip_file_path = "C:\\2007-2016-Homelessnewss-USA.csv\\homelessness.zip"

# Define the directory to extract the files to
extract_dir = "7-2016-Homelessnewss-USA.csv"  # Replace with the directory where you want to extract the files

# Unzip the downloaded file
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_dir)

# Remove the downloaded zip file
os.remove(zip_file_path)






In [163]:
df2 = pd.read_csv('7-2016-Homelessnewss-USA.csv/2007-2016-Homelessnewss-USA.csv')

In [297]:
df.head()

brokered_by    status     price  bed  bath  acre_lot     street  \
2270      94933.0  for_sale  950000.0  5.0   4.0      0.99  1260473.0   
2493      22968.0  for_sale   79500.0  NaN   NaN      0.76   765975.0   
2555      34543.0  for_sale  176500.0  NaN   NaN      0.50   917409.0   
3062      22968.0  for_sale  129000.0  NaN   NaN      6.30   739696.0   
3409      21163.0  for_sale  525000.0  3.0   3.0      0.45  1813270.0   

               city           state  zip_code  house_size       Year  
2270   Saint Thomas  Virgin Islands     802.0      5000.0 2013-10-11  
2493  Christiansted  Virgin Islands     820.0         NaN 2016-04-28  
2555  Christiansted  Virgin Islands     820.0         NaN 2016-11-16  
3062   Frederiksted  Virgin Islands     840.0         NaN 2010-01-26  
3409         Agawam   Massachusetts    1001.0      2314.0 2014-06-25

In [299]:
df.tail(10)

brokered_by    status      price  bed  bath  acre_lot     street  \
1414163       3571.0  for_sale   115000.0  NaN   NaN      0.19   549588.0   
1414166      48892.0  for_sale   149900.0  NaN   NaN      0.83   930101.0   
1414186       6482.0  for_sale   859900.0  5.0   3.0      0.40  1318135.0   
1414187      79471.0  for_sale   439900.0  3.0   2.0      0.09  1251298.0   
1414188      56985.0  for_sale   515000.0  4.0   2.0      0.23  1773566.0   
1414189      79471.0  for_sale   425000.0  3.0   2.0      0.16  1728797.0   
1414192      95190.0  for_sale   414900.0  4.0   2.0      0.23  1706555.0   
1414316      21688.0  for_sale   525000.0  3.0   2.0      0.46  1003563.0   
1414327      57432.0  for_sale  1100000.0  4.0   4.0      0.32  1280711.0   
1414346      82020.0  for_sale   497000.0  4.0   3.0      0.41  1698210.0   

              city   state  zip_code  house_size       Year  
1414163     Juneau  Alaska   99801.0         NaN 2015-08-20  
1414166     Juneau  Alaska   99801.0         NaN 2015-06-22  
1414186     Juneau  Alaska   99801.0      3689.0 2010-06-01  
1414187     Juneau  Alaska   99801.0      1770.0 2016-08-15  
1414188     Juneau  Alaska   99801.0      1894.0 2008-09-15  
1414189     Juneau  Alaska   99801.0      1192.0 2015-12-18  
1414192     Juneau  Alaska   99801.0      1304.0 2013-03-26  
1414316  Ketchikan  Alaska   99901.0      1272.0 2007-06-12  
1414327  Ketchikan  Alaska   99901.0      3600.0 2011-09-09  
1414346  Ketchikan  Alaska   99901.0      2676.0 2014-03-07

In [300]:
df.describe() 

brokered_by         price            bed           bath  \
count  199118.000000  1.992690e+05  177895.000000  178740.000000   
mean    52699.658660  6.368690e+05       3.291442       2.564860   
min         0.000000  1.000000e+00       1.000000       1.000000   
25%     23592.000000  2.150000e+05       3.000000       2.000000   
50%     52770.000000  3.700000e+05       3.000000       2.000000   
75%     79221.000000  6.350000e+05       4.000000       3.000000   
max    110142.000000  2.500000e+08     123.000000     123.000000   
std     30804.528234  1.728620e+06       1.450949       1.419913   

            acre_lot        street       zip_code    house_size  \
count  169610.000000  1.984410e+05  199268.000000  1.689570e+05   
mean        9.126097  9.401535e+05   51137.008376  2.188819e+03   
min         0.000000  7.300000e+01     802.000000  4.000000e+00   
25%         0.150000  4.732520e+05   29605.000000  1.296000e+03   
50%         0.230000  9.400200e+05   46074.000000  1.784000e+03   
75%         0.570000  1.403475e+06   77469.000000  2.525000e+03   
max    100000.000000  2.001338e+06   99901.000000  3.434706e+06   
std       604.058117  5.392514e+05   28654.532876  9.993093e+03   

                                Year  
count                         199269  
mean   2012-10-14 01:57:37.892597760  
min              2007-01-01 00:00:00  
25%              2010-03-19 00:00:00  
50%              2013-05-31 00:00:00  
75%              2015-06-24 00:00:00  
max              2016-12-30 00:00:00  
std                              NaN

In [301]:
df.columns

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'Year'],
      dtype='object')

In [302]:
!pip install kaggle

In [306]:
print(df.isnull().sum())


brokered_by      151
status             0
price              0
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


In [307]:
print(df.duplicated().sum())


0


In [314]:
#FILL IN 1st data gathering and loading method

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**

"dirty data and messy data.

Inaccurate data is an example of dirty data, i.e., a data quality issue. High data quality is crucial for obtaining reliable insights and driving informed decision-makin.

Unorganized data is an example of messy data, i.e., a data structural issue. With tidy data, we can focus on performing additional assessments and cleaning without dealing with roadblocks around getting the data into a structure that's easier to view and pa" 
 data.rse.

In [317]:
df2.head(10)

Year State CoC Number       CoC Name  \
0 2007-01-01    AK     AK-500  Anchorage CoC   
1 2007-01-01    AK     AK-500  Anchorage CoC   
2 2007-01-01    AK     AK-500  Anchorage CoC   
3 2007-01-01    AK     AK-500  Anchorage CoC   
4 2007-01-01    AK     AK-500  Anchorage CoC   
5 2007-01-01    AK     AK-500  Anchorage CoC   
6 2007-01-01    AK     AK-500  Anchorage CoC   
7 2007-01-01    AK     AK-500  Anchorage CoC   
8 2007-01-01    AK     AK-500  Anchorage CoC   
9 2007-01-01    AK     AK-500  Anchorage CoC   

                                       Measures  Count  
0              Chronically Homeless Individuals    224  
1                          Homeless Individuals    696  
2                   Homeless People in Families    278  
3    Sheltered Chronically Homeless Individuals    187  
4                            Sheltered Homeless    842  
5                Sheltered Homeless Individuals    589  
6         Sheltered Homeless People in Families    253  
7                                Total Homeless    974  
8  Unsheltered Chronically Homeless Individuals     37  
9                          Unsheltered Homeless    132

In [319]:
df2.tail(10)

Year State CoC Number               CoC Name  \
86519 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86520 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86521 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86522 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86523 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86524 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86525 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86526 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86527 2016-01-01    WY     WY-500  Wyoming Statewide CoC   
86528 2016-01-01    WY     WY-500  Wyoming Statewide CoC   

                                                Measures  Count  
86519                               Unsheltered Homeless    366  
86520                   Unsheltered Homeless Individuals    240  
86521            Unsheltered Homeless People in Families    126  
86522  Unsheltered Homeless Unaccompanied Children (U...      2  
86523  Unsheltered Homeless Unaccompanied Young Adult...      5  
86524  Unsheltered Homeless Unaccompanied Youth (Unde...      7  
86525                      Unsheltered Homeless Veterans     31  
86526             Unsheltered Parenting Youth (Under 25)      3  
86527              Unsheltered Parenting Youth Age 18-24      3  
86528               Unsheltered Parenting Youth Under 18      0

In [321]:
df2.describe()

Year         Count
count                          86529  86529.000000
mean   2012-10-14 07:01:55.202995712    332.286748
min              2007-01-01 00:00:00      0.000000
25%              2011-01-01 00:00:00      7.000000
50%              2013-01-01 00:00:00     52.000000
75%              2015-01-01 00:00:00    235.000000
max              2016-01-01 00:00:00  75323.000000
std                              NaN   1492.614929

In [323]:
print(df2.isnull().sum())


Year          0
State         0
CoC Number    0
CoC Name      0
Measures      0
Count         0
dtype: int64


In [331]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Year        86529 non-null  datetime64[ns]
 1   State       86529 non-null  object        
 2   CoC Number  86529 non-null  object        
 3   CoC Name    86529 non-null  object        
 4   Measures    86529 non-null  object        
 5   Count       86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(4)
memory usage: 3.6+ MB


In [332]:
df2.columns

Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', 'Count'], dtype='object')

In [334]:
df2.value_counts()

Year        State  CoC Number  CoC Name                                 Measures                                           Count
2007-01-01  AK     AK-500      Anchorage CoC                            Chronically Homeless Individuals                   224      1
2015-01-01  IL     IL-504      Madison County CoC                       Homeless Unaccompanied Youth (Under 25)            1        1
                                                                        Sheltered Chronically Homeless People in Families  10       1
                                                                        Sheltered Chronically Homeless Individuals         5        1
                                                                        Sheltered Chronically Homeless                     15       1
                                                                                                                                   ..
2012-01-01  CO     CO-503      Metropolitan Denver Homeless Initiat

In [336]:
#Optional: store the raw data in your local data store


> data quality:

  * *Completeness: is a metric that helps you understand whether your data is sufficient to answer interesting questions or solve your problem*

 *  *Validity is a metric helping you understand how well your data conforms to a defined set of rules for data, also known as a schema* 
   
* Accuracy is a metric that helps you understand whether your data accurately represents the reality it aims to depict.* 
* Consistency is a metric that helps you understand two things: whether your data follows a standard format and whether your data’s info matches with information from other data sources.
* Uniqueness is a metric that helps you understand whether there are duplicate or overlapping values in your data.

### Quality Issue and Tidiness Issues (Notes for myself)

#FILL IN - Inspecting the data frame visually 

*  The shape of the df is huge It covers a lot of years, but df2 covers from 2007 to 2016: *So We have to limit the shape according to year. 
**( drop the data after 2017 | 12/2016  so it fits with the second dataset, or Filter the DataFrame to include only data from 2007 to 2016)** 
> And the code for this is: 
 df = df[(df['Year'] >= 2007) & (df['Year'] <= 2016)]   
- Year here is (prev_sold_date) So it have to be renamed either ways to ["Year"] which is object have to be changed to datetime.
- The prev_sold_date  object, and it must be datatime. And it's important to notice that this column gives us an information saying this the "Sold date", So I have to use this information correctly to connect the 2 datasets, it's important to notice it might be a key column instead of the state.
-  I have in df['prev_sold_date'] a 734297 null value.. first, I will reshape the data, then decide on how to deal with 
    date null values. 

* In df2 "count" must be integers.

* df Have a lot of nulls. However, the most needed ones is the prices and the state. And df2 does not have any nulls.
* To Fix this issue, I must define the necessary columns in both datas. 
  - necessary_columnsdf = df['Price', 'State', 'sold_date/ Year']
  - necessary_columnsdf2 = df2['State', 'Year', 'Measures', 'counts']
* * df have 8 missing Values in ['State'] I have to change it to ['unknown'] or fill it manually by knowing the city. 

* In df2 ["States"] must be defined by name So I have to change it be matches df1.

  > df2['State'] = df2['State'].replace({'Alaska':'AK', 'Alabama':'AL', 'Arkansas':'AR', 'Arizona':'AZ', 
                      'California':'CA', 'Colorado':'CO', 'Connecticut':'CT', 
                      'District of Columbia':'DC', 'Delaware':'DE', 'Florida':'FL', 
                      'Georgia':'GA', 'Hawaii':'HI', 'Iowa':'IA', 
                      'Idaho':'ID', 'Illinois':'IL', 'Indiana':'IN', 'Kansas':'KS', 
                      'Kentucky':'KY', 'Louisiana':'LA', 'Massachusetts':'MA', 'Maryland':'MD', 
                      'Maine':'ME', 'Michigan':'MI', 'Minnesota':'MN', 'Missouri':'MO', 
                      'Mississippi':'MS', 'Montana':'MT', 'North Carolina':'NC', 
                      'North Dakota':'ND', 'Nebraska':'NE', 'New Hampshire':'NH', 
                      'New Jersey':'NJ', 'New Mexico':'NM', 'Nevada':'NV', 'New York':'NY', 
                      'Ohio':'OH', 'Oklahoma':'OK', 'Oregon':'OR', 'Pennsylvania':'PA', 
                      'Puerto Rico':'PR', 'Rhode Island':'RI', 'South Carolina':'SC', 
                      'South Dakota':'SD', 'Tennessee':'TN', 'Texas':'TX', 'Utah':'UT', 
                      'Virginia':'VA', 'Vermont':'VT', 'Washington':'WA', 
                      'Wisconsin':'WI', 'West Virginia':'WV', 'Wyoming':'WY'})

* In df2 Measures addresses many kind of "Homelessness", which are Homeless Individuals',' Homeless People in Families', 'Unsheltered Homeless', We can make it moinge Tidy make it i \\\\\ also there is " Unsheltered Parenting Youth Under 18" And "...Under 25.." n one category
   > hodf2ols = homeless['Measures'].str.df2ies()
homeless.join(homeless_cols[['Homeless Individuals','Homeless People in Families', 'Unshelter
     - Homeless']]*
* But still,I dont think we have to make it in categories.-  Instead, I Think I have to collect the "count" number in one re
  ult per state. 
 bject

In [342]:
df.shape

(199269, 12)

In [344]:
df2.shape

(86529, 6)

In [346]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 199269 entries, 2270 to 1414346
Data columns (total 12 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   brokered_by  199118 non-null  float64       
 1   status       199269 non-null  object        
 2   price        199269 non-null  float64       
 3   bed          177895 non-null  float64       
 4   bath         178740 non-null  float64       
 5   acre_lot     169610 non-null  float64       
 6   street       198441 non-null  float64       
 7   city         199251 non-null  object        
 8   state        199269 non-null  object        
 9   zip_code     199268 non-null  float64       
 10  house_size   168957 non-null  float64       
 11  Year         199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(8), object(3)
memory usage: 19.8+ MB


In [348]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Year        86529 non-null  datetime64[ns]
 1   State       86529 non-null  object        
 2   CoC Number  86529 non-null  object        
 3   CoC Name    86529 non-null  object        
 4   Measures    86529 non-null  object        
 5   Count       86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(4)
memory usage: 3.6+ MB


In [350]:
df.sample(5)

brokered_by    status      price  bed  bath  acre_lot     street  \
1144741      78251.0  for_sale   918888.0  4.0   2.0     10.01   224744.0   
172061       22561.0  for_sale    79900.0  3.0   1.0      0.12   169948.0   
389900        5660.0  for_sale  1650000.0  6.0   6.0      0.62   965116.0   
1027485      53465.0  for_sale   324900.0  3.0   2.0      0.31  1058017.0   
1038721      57305.0  for_sale    45000.0  NaN   NaN      0.30  1773810.0   

               city     state  zip_code  house_size       Year  
1144741    Larkspur  Colorado   80118.0      4457.0 2014-02-14  
172061   Cumberland  Maryland   21502.0      1080.0 2015-03-25  
389900      Atlanta   Georgia   30319.0      4700.0 2016-06-17  
1027485   Arlington     Texas   76002.0      1579.0 2007-03-27  
1038721  Brookeland     Texas   75931.0         NaN 2014-03-27

In [351]:
df.columns

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'Year'],
      dtype='object')

In [354]:
df2.sample(5)

Year State CoC Number  \
81878 2016-01-01    NY     NY-604   
66570 2015-01-01    PA     PA-510   
3342  2007-01-01    NY     NY-512   
85596 2016-01-01    VA     VA-601   
86281 2016-01-01    WI     WI-503   

                                                CoC Name  \
81878  Yonkers/Mount Vernon/New Rochelle/Westchester CoC   
66570                        Lancaster City & County CoC   
3342                          Troy/Rensselaer County CoC   
85596                                 Fairfax County CoC   
86281                            Madison/Dane County CoC   

                                    Measures  Count  
81878              Parenting Youth Age 18-24     99  
66570       Unsheltered Homeless Individuals     10  
3342   Sheltered Homeless People in Families    125  
85596       Unsheltered Homeless Individuals     74  
86281                   Homeless Individuals    360

In [355]:
print(df.isnull().sum())

brokered_by      151
status             0
price              0
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


In [356]:
print(df.columns)

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'Year'],
      dtype='object')


In [357]:
print(df2.isnull().sum())

Year          0
State         0
CoC Number    0
CoC Name      0
Measures      0
Count         0
dtype: int64


In [362]:
df2['Measures'].unique()

array(['Chronically Homeless Individuals', 'Homeless Individuals',
       'Homeless People in Families',
       'Sheltered Chronically Homeless Individuals', 'Sheltered Homeless',
       'Sheltered Homeless Individuals',
       'Sheltered Homeless People in Families', 'Total Homeless',
       'Unsheltered Chronically Homeless Individuals',
       'Unsheltered Homeless', 'Unsheltered Homeless Individuals',
       'Unsheltered Homeless People in Families', 'Chronically Homeless',
       'Chronically Homeless People in Families', 'Homeless Veterans',
       'Sheltered Chronically Homeless',
       'Sheltered Chronically Homeless People in Families',
       'Sheltered Homeless Veterans', 'Unsheltered Chronically Homeless',
       'Unsheltered Chronically Homeless People in Families',
       'Unsheltered Homeless Veterans', 'Children of Parenting Youth',
       'Homeless Unaccompanied Children (Under 18)',
       'Homeless Unaccompanied Young Adults (Age 18-24)',
       'Homeless Unaccompan

In [364]:
df2['Measures'].nunique()

42

In [366]:
df2['Measures']. duplicated().sum()

86487

In [368]:
df2['Measures']. isnull().sum()

0

In [370]:
df2['Measures'].value_counts

<bound method IndexOpsMixin.value_counts of 0                         Chronically Homeless Individuals
1                                     Homeless Individuals
2                              Homeless People in Families
3               Sheltered Chronically Homeless Individuals
4                                       Sheltered Homeless
                               ...                        
86524    Unsheltered Homeless Unaccompanied Youth (Unde...
86525                        Unsheltered Homeless Veterans
86526               Unsheltered Parenting Youth (Under 25)
86527                Unsheltered Parenting Youth Age 18-24
86528                 Unsheltered Parenting Youth Under 18
Name: Measures, Length: 86529, dtype: object>

## 3. Clean data
Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

# For the porpuse of combine 

Here is to make the time zone for df the same as df2: 
- Rename prev_sold_date To "Year". 
- dtype: datetime
- 1/1/2006 - 31/12/2017 

In [375]:
df.rename(columns= {'prev_sold_date': 'Year'}, inplace= True)  

Check: 

In [378]:
df.sample(1)

brokered_by    status     price  bed  bath  acre_lot    street  \
705391      45814.0  for_sale  440000.0  4.0   3.0      0.34  786391.0   

             city state  zip_code  house_size       Year  
705391  Beachwood  Ohio   44122.0      2623.0 2016-11-29

Change the dtype for Year because it's object

In [381]:
df['Year'] = pd.to_datetime(df['Year'], errors='coerce')

Check

In [384]:
df.dtypes

brokered_by           float64
status                 object
price                 float64
bed                   float64
bath                  float64
acre_lot              float64
street                float64
city                   object
state                  object
zip_code              float64
house_size            float64
Year           datetime64[ns]
dtype: object

In [386]:
df['Year']. isnull().sum()

0

Now, lemit the df dataframe

In [389]:
df = df[(df['Year'].dt.year >= 2007) & (df['Year'].dt.year <= 2016)]


Check, The shape of df was (2226382, 12)

In [392]:
df.shape

(199269, 12)

In [394]:
df['Year']. isnull().sum()

0

In [396]:
check = 2226382 - 199569
print(check)

2026813


So, I've dropped 2026813 intries. I have to check the nulls. 

In [399]:
print(df.isnull().sum())

brokered_by      151
status             0
price              0
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


Now,
- I have no nulls in the Year Column. 
- I have no nulls in the State Column
- I have 69 in the Price column. I can fill it with th mean of every year to care about the accuracy. 

### **Quality Issue 1: FILL IN**

In [415]:
# FILL IN - Validate the cleaning was successful
# Here,  I must fill the Na in prices with the mean of the corresponding year.

# Calculate the mean price for each year
mean_prices = df.groupby('Year')["price"].transform('mean')

# Fill missing prices with the mean price of the corresponding year
df["price"] = df["price"].fillna(mean_prices)

# Display the DataFrame after filling missing prices
print("DataFrame after filling missing prices with the mean price of each year:")
print(df)


DataFrame after filling missing prices with the mean price of each year:
         brokered_by    status      price  bed  bath  acre_lot     street  \
2270         94933.0  for_sale   950000.0  5.0   4.0      0.99  1260473.0   
2493         22968.0  for_sale    79500.0  NaN   NaN      0.76   765975.0   
2555         34543.0  for_sale   176500.0  NaN   NaN      0.50   917409.0   
3062         22968.0  for_sale   129000.0  NaN   NaN      6.30   739696.0   
3409         21163.0  for_sale   525000.0  3.0   3.0      0.45  1813270.0   
...              ...       ...        ...  ...   ...       ...        ...   
1414189      79471.0  for_sale   425000.0  3.0   2.0      0.16  1728797.0   
1414192      95190.0  for_sale   414900.0  4.0   2.0      0.23  1706555.0   
1414316      21688.0  for_sale   525000.0  3.0   2.0      0.46  1003563.0   
1414327      57432.0  for_sale  1100000.0  4.0   4.0      0.32  1280711.0   
1414346      82020.0  for_sale   497000.0  4.0   3.0      0.41  1698210.0   

  

In [417]:
print(df['price'])

2270        950000.0
2493         79500.0
2555        176500.0
3062        129000.0
3409        525000.0
             ...    
1414189     425000.0
1414192     414900.0
1414316     525000.0
1414327    1100000.0
1414346     497000.0
Name: price, Length: 199269, dtype: float64


Check: 

In [420]:
print(df.isnull().sum())

brokered_by      151
status             0
price              0
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [424]:
df2.columns

Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', 'Count'], dtype='object')

In [430]:
#FILL IN - Apply the cleaning strategy
# In df2 "count" must be integers
df2['Count'] = df2['Count'].str.replace(',', '')

df2['Count'] = df2['Count'].astype(int)

AttributeError: Can only use .str accessor with string values!

Check

In [433]:
#FILL IN - Validate the cleaning was successful
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Year        86529 non-null  datetime64[ns]
 1   State       86529 non-null  object        
 2   CoC Number  86529 non-null  object        
 3   CoC Name    86529 non-null  object        
 4   Measures    86529 non-null  object        
 5   Count       86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(4)
memory usage: 3.6+ MB


Now, The Years Column is object here too, and it must be datetime too. 

In [436]:
df2['Year'] = pd.to_datetime(df2['Year'])

In [438]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Year        86529 non-null  datetime64[ns]
 1   State       86529 non-null  object        
 2   CoC Number  86529 non-null  object        
 3   CoC Name    86529 non-null  object        
 4   Measures    86529 non-null  object        
 5   Count       86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(4)
memory usage: 3.6+ MB


Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [444]:
#FILL IN - Apply the cleaning strategy
# Here I have to change the state column so when I combine the data make sense. 

state_abbr_to_full = {
    'AK': 'Alaska', 'AL': 'Alabama', 'AR': 'Arkansas', 'AZ': 'Arizona', 'CA': 'California',
    'CO': 'Colorado', 'CT': 'Connecticut', 'DC': 'District of Columbia', 'DE': 'Delaware', 'FL': 'Florida',
    'GA': 'Georgia', 'HI': 'Hawaii', 'IA': 'Iowa', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana',
    'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'MA': 'Massachusetts', 'MD': 'Maryland',
    'ME': 'Maine', 'MI': 'Michigan', 'MN': 'Minnesota', 'MO': 'Missouri', 'MS': 'Mississippi', 'MT': 'Montana',
    'NC': 'North Carolina', 'ND': 'North Dakota', 'NE': 'Nebraska', 'NH': 'New Hampshire', 'NJ': 'New Jersey',
    'NM': 'New Mexico', 'NV': 'Nevada', 'NY': 'New York', 'OH': 'Ohio', 'OK': 'Oklahoma', 'OR': 'Oregon',
    'PA': 'Pennsylvania', 'PR': 'Puerto Rico', 'RI': 'Rhode Island', 'SC': 'South Carolina', 'SD': 'South Dakota',
    'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah', 'VA': 'Virginia', 'VT': 'Vermont', 'WA': 'Washington',
    'WI': 'Wisconsin', 'WV': 'West Virginia', 'WY': 'Wyoming', 'VI': 'Virgin Islands', 'GU': 'Guam'
}

# Replace state abbreviations with full names
df2['State'] = df2['State'].replace(state_abbr_to_full)

#States in df2 but not in df:
#{'Puerto Rico', 'VI', 'GU'}

#States in df but not in df2:
#{'Virgin Islands'}

In [450]:
print(df2['State'].unique())

['Alaska' 'Alabama' 'Arkansas' 'Arizona' 'California' 'Colorado'
 'Connecticut' 'District of Columbia' 'Delaware' 'Florida' 'Georgia'
 'Guam' 'Hawaii' 'Iowa' 'Idaho' 'Illinois' 'Indiana' 'Kansas' 'Kentucky'
 'Louisiana' 'Massachusetts' 'Maryland' 'Maine' 'Michigan' 'Minnesota'
 'Missouri' 'Mississippi' 'Montana' 'North Carolina' 'North Dakota'
 'Nebraska' 'New Hampshire' 'New Jersey' 'New Mexico' 'Nevada' 'New York'
 'Ohio' 'Oklahoma' 'Oregon' 'Pennsylvania' 'Puerto Rico' 'Rhode Island'
 'South Carolina' 'South Dakota' 'Tennessee' 'Texas' 'Utah' 'Virginia'
 'Virgin Islands' 'Vermont' 'Washington' 'Wisconsin' 'West Virginia'
 'Wyoming']


In [460]:
df2.sample(5)

Year          State CoC Number  \
1421  2007-01-01       Illinois     IL-508   
81122 2016-01-01       New York     NY-505   
76742 2016-01-01  Massachusetts     MA-519   
14478 2010-01-01        Alabama     AL-507   
30624 2012-01-01      Louisiana     LA-507   

                                                CoC Name  \
1421   East Saint Louis/Belleville/Saint Clair County...   
81122  Syracuse, Auburn/Onondaga, Oswego, Cayuga Coun...   
76742               Attleboro/Taunton/Bristol County CoC   
14478                       Alabama Balance of State CoC   
30624                   Alexandria/Central Louisiana CoC   

                                    Measures  Count  
1421          Sheltered Homeless Individuals    146  
81122              Parenting Youth Age 18-24     36  
76742   Unsheltered Parenting Youth Under 18      0  
14478  Sheltered Homeless People in Families    104  
30624       Unsheltered Homeless Individuals     15

In [462]:
df2['State'].nunique()

54

In [464]:
df['state']. nunique()

52

In [466]:
df.duplicated(). sum()

0

In [468]:
df2.duplicated(). sum()

0

I have 2 more issues 
> 
1- The state is in df2 in capital and in df is small. Must change to capital 
2- There is 2 more "State" in df2. 

In [471]:
df.rename(columns={'state': 'State'}, inplace=True)


In [473]:
df.head()

brokered_by    status     price  bed  bath  acre_lot     street  \
2270      94933.0  for_sale  950000.0  5.0   4.0      0.99  1260473.0   
2493      22968.0  for_sale   79500.0  NaN   NaN      0.76   765975.0   
2555      34543.0  for_sale  176500.0  NaN   NaN      0.50   917409.0   
3062      22968.0  for_sale  129000.0  NaN   NaN      6.30   739696.0   
3409      21163.0  for_sale  525000.0  3.0   3.0      0.45  1813270.0   

               city           State  zip_code  house_size       Year  
2270   Saint Thomas  Virgin Islands     802.0      5000.0 2013-10-11  
2493  Christiansted  Virgin Islands     820.0         NaN 2016-04-28  
2555  Christiansted  Virgin Islands     820.0         NaN 2016-11-16  
3062   Frederiksted  Virgin Islands     840.0         NaN 2010-01-26  
3409         Agawam   Massachusetts    1001.0      2314.0 2014-06-25

In [475]:

# Get the unique states in both DataFrames
unique_states_df = set(df['State'].unique())
unique_states_df2 = set(df2['State'].unique())

# Find states that are in df2 but not in df
states_in_df2_not_in_df = unique_states_df2 - unique_states_df

# Find states that are in df but not in df2
states_in_df_not_in_df2 = unique_states_df - unique_states_df2

print("States in df2 but not in df:")
print(states_in_df2_not_in_df)

print("\nStates in df but not in df2:")
print(states_in_df_not_in_df2)




States in df2 but not in df:
{'Guam', 'Puerto Rico'}

States in df but not in df2:
set()


#### So, I must drop States in df2 but not in df: 
{'Puerto Rico', 'Guam' 

- Now, I want to check the year column for tidiness purposes. I have noticed that the entries of the Year column are not tidy in df but they are tidy in df2.}

In [478]:
df =df.sort_values(by='Year').reset_index()

In [480]:
df.head(5)

index  brokered_by    status     price  bed  bath  acre_lot     street  \
0   911844      15977.0  for_sale   80000.0  NaN   NaN      3.00   659959.0   
1   864685       8748.0  for_sale  541000.0  4.0   3.0      0.25   811025.0   
2   829710      22490.0  for_sale  259000.0  2.0   3.0      0.25    77733.0   
3   864687       8748.0  for_sale  491245.0  4.0   3.0      0.25   811057.0   
4  1138769      44642.0  for_sale  410000.0  NaN   NaN      0.11  1253561.0   

           city      State  zip_code  house_size       Year  
0        Mexico   Missouri   65265.0         NaN 2007-01-01  
1    Plainfield   Illinois   60586.0      3084.0 2007-01-01  
2  Grand Rapids  Minnesota   55744.0      2360.0 2007-01-01  
3    Plainfield   Illinois   60586.0      2612.0 2007-01-01  
4      Superior   Colorado   80027.0         NaN 2007-01-02

In [482]:
df2.sample(5)

Year       State CoC Number  \
80978 2016-01-01    New York     NY-500   
72569 2016-01-01    Delaware     DE-500   
36258 2013-01-01    Arkansas     AR-512   
8349  2008-01-01        Ohio     OH-502   
70887 2016-01-01  California     CA-509   

                                             CoC Name  \
80978  Rochester/Irondequoit/Greece/Monroe County CoC   
72569                          Delaware Statewide CoC   
36258      Boone, Baxter, Marion, Newton Counties CoC   
8349                    Cleveland/Cuyahoga County CoC   
70887                            Mendocino County CoC   

                                                Measures  Count  
80978  Unsheltered Homeless Unaccompanied Children (U...      1  
72569                 Sheltered Parenting Youth Under 18      0  
36258                               Homeless Individuals     41  
8349                                Unsheltered Homeless    151  
70887               Sheltered Parenting Youth (Under 25)      2

# Measures 
*I have 42 kinds about Homeless people, So I will make them in 2 categories*

1- Sheltered 
2- Unsheltered.
However, I Still will have ['Chronically Homeless'] and ['Total Homeless'] not recognized. I will consider "Chronically Homeless" as Unsheltered Homeless Because "Chronically" must mean that. 
And, Total homeless means both sheltered and unsheltered I will consider the number in both sheltered and unsheltered. 


In [485]:
# Define keywords for categorization
sheltered_keywords = ['Sheltered', 'Parenting Youth', 'Chronically Homeless', 'Total Homeless', 'Homeless People in Families'] 
unsheltered_keywords = ['Unsheltered', 'Chronically Homeless', 'Homeless Unaccompanied', 'Total Homeless', 'Homeless Individuals']

# Categorize measures into 'Sheltered' and 'Unsheltered'
df2.loc[:, 'Sheltered'] = df2['Measures'].apply(lambda x: any(keyword in x for keyword in sheltered_keywords))
df2.loc[:, 'Unsheltered'] = df2['Measures'].apply(lambda x: any(keyword in x for keyword in unsheltered_keywords))


# Display DataFrame
print(df2)

            Year    State CoC Number               CoC Name  \
0     2007-01-01   Alaska     AK-500          Anchorage CoC   
1     2007-01-01   Alaska     AK-500          Anchorage CoC   
2     2007-01-01   Alaska     AK-500          Anchorage CoC   
3     2007-01-01   Alaska     AK-500          Anchorage CoC   
4     2007-01-01   Alaska     AK-500          Anchorage CoC   
...          ...      ...        ...                    ...   
86524 2016-01-01  Wyoming     WY-500  Wyoming Statewide CoC   
86525 2016-01-01  Wyoming     WY-500  Wyoming Statewide CoC   
86526 2016-01-01  Wyoming     WY-500  Wyoming Statewide CoC   
86527 2016-01-01  Wyoming     WY-500  Wyoming Statewide CoC   
86528 2016-01-01  Wyoming     WY-500  Wyoming Statewide CoC   

                                                Measures  Count  Sheltered  \
0                       Chronically Homeless Individuals    224       True   
1                                   Homeless Individuals    696      False   
2        

In [492]:
df2.columns

Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', 'Count',
       'Sheltered', 'Unsheltered'],
      dtype='object')

I have To drop Measures, After I make a copy.
**Drop the 'Measures' column**
df2 = df2.drop(columns=['Measures'])

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [496]:
#FILL IN - Remove unnecessary variables and combine datasets
# necessary_columnsdf = df['Price', 'State', 'Year']
# necessary_columnsdf2 = df2['State', 'Year', 'Measures', 'counts']

necessary_df = df[['price','State', 'Year']]
necessary_df2 = df2 [['State', 'Year', 'Measures', 'Count']]

df = necessary_df
df2 = necessary_df2



Check

In [499]:
df.head(2)

price     State       Year
0   80000.0  Missouri 2007-01-01
1  541000.0  Illinois 2007-01-01

In [501]:
df2.head(2)

State       Year                          Measures  Count
0  Alaska 2007-01-01  Chronically Homeless Individuals    224
1  Alaska 2007-01-01              Homeless Individuals    696

In [503]:
df2.columns

Index(['State', 'Year', 'Measures', 'Count'], dtype='object')

In [505]:
df.columns

Index(['price', 'State', 'Year'], dtype='object')

In [507]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   State     86529 non-null  object        
 1   Year      86529 non-null  datetime64[ns]
 2   Measures  86529 non-null  object        
 3   Count     86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(2)
memory usage: 2.3+ MB


In [509]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float64       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 4.6+ MB


- So, I must drop States in df2 but not in df: 
{'Puerto Rico', 'Guam}'

In [512]:
# States in df but not in df2
states_to_drop = {'Puerto Rico', 'Guam'}

# Drop rows in df2 where the state is in states_to_drop
df2 = df2[~df2['State'].isin(states_to_drop)]

In [514]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   State     85881 non-null  object        
 1   Year      85881 non-null  datetime64[ns]
 2   Measures  85881 non-null  object        
 3   Count     85881 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(2)
memory usage: 2.9+ MB


## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

# DF 2 - Homelessness in USA

In [518]:
# Save raw data to CSV
df2.to_csv('raw_homeless_data.csv', index=False)

In [520]:
df2 = df2.drop(columns=['Measures']) #drop the Measures data

In [522]:
df2.head()

State       Year  Count
0  Alaska 2007-01-01    224
1  Alaska 2007-01-01    696
2  Alaska 2007-01-01    278
3  Alaska 2007-01-01    187
4  Alaska 2007-01-01    842

In [524]:
df2.to_csv('cleaned_homeless_data.csv', index=False)

print("Raw and cleaned data have been saved to CSV files.")


Raw and cleaned data have been saved to CSV files.


# DF 1 : Real Estate in USA 

In [527]:
df.to_csv('cleaned_Houses_MArket_USA_data.csv', index=False)

In [528]:
df.head(2)

price     State       Year
0   80000.0  Missouri 2007-01-01
1  541000.0  Illinois 2007-01-01

In [531]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float64       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 4.6+ MB


In [ ]:
df. drop('street', axis=1)

In [533]:
df.loc[:, 'price'] = df['price'].astype(np.float32)

In [535]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float64       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 4.6+ MB


In [537]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   State   85881 non-null  object        
 1   Year    85881 non-null  datetime64[ns]
 2   Count   85881 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(1)
memory usage: 2.3+ MB


In [539]:
df2.loc[:, 'Count'] = df2['Count'].astype(np.int16)

In [541]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   State   85881 non-null  object        
 1   Year    85881 non-null  datetime64[ns]
 2   Count   85881 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(1)
memory usage: 2.3+ MB


# Combine 2 datasets

In [ ]:
combined_df = pd.merge(df, df2, on='State', how='inner')

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN